clang/lib/Serialization/ASTReader.cpp

```diff
@@ -5834,9 +5834,13 @@ bool ASTReader::readASTFileControlBlock(
           break;
         case INPUT_FILE:
           bool Overridden = static_cast<bool>(Record[3]);
+          size_t FilenameLen = ModuleDir.size() + Record[7] + 1;
           auto Filename = ResolveImportedPath(PathBuf, Blob, ModuleDir);
+          StringRef FilenameAsRequested = Filename->substr(0, FilenameLen);
+          StringRef ExternalFilename = Filename->substr(FilenameLen);
```

Comment:
Can you explain how this is supposed to work? IIRC `Blob` is the concatenation of the as-requested name and the name, so this doesn't make sense to me. (To be frank, even the original code seems incorrect.)

Reply:
> IIRC Blob is the concatenation of the as-requested name and the name. 

That is correct. 

Basically `Record[7]` encodes the length of the basename (or filename without the full path) of the vfs version of the file.  https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ASTWriter.cpp#L1902

`ModuleDir` is the directory of the vfs-version of the directory those vfs mapped files live in.  https://github.com/llvm/llvm-project/pull/132237/files#diff-c61a3cce4bfa099b5af032fa83cbf1563f0af4bf58dc112b39571d74b6b681c1R5784 

With that combination I derived the full length of of the vfs file path. The external path is the rest.

Reply:
Ok, got it.

What's confusing to me is that `Filename` is essentially `"<ModuleDir>/<FilenameAsRequested><Filename>"`, which doesn't mean anything. And `ExternalFilename` is just `<Filename>`, which is unresolved.

Why don't we split `Blob` like we do elsewhere (`Blob.substr(0, AsRequestedLength)` and `Blob.substr(AsRequestedLength)`) and call `ResolveImportedPath()` on both?

Reply:
Applied your suggestion except for calling `ResolveImportedPath` on both parts of the substring. Let me know if I am missing something, but the external path should always be absolute, so I think it would be unnecessary. It also reduces some logic of needing to copy strings since  

> The return value must go out of scope before the next call to \c ResolveImportedPath.

Reply:
Both of these go through `PreparePathForOutput()` in `ASTWriter`, so both of these can be relative.